lib/ui/channel_buffers.dart

```diff
@@ -0,0 +1,187 @@
+      _queue.addLast(val);
+      return overflowCount > 0;
+    }
+  }
+
+  /// Returns null when empty.
+  T pop() {
+    return _queue.isEmpty ? null : _queue.removeFirst();
+  }
+
+  /// Removes items until then length reaches [lengthLimit] and returns
+  /// the number of items removed.
+  int _dropOverflowItems(int lengthLimit) {
+    int result = 0;
+    while (_queue.length > lengthLimit) {
+      final T item = _queue.removeFirst();
+      if (_dropItemCallback != null) {
+        _dropItemCallback(item);
+      }
+      result += 1;
+    }
+    return result;
+  }
+
+  /// Returns the number of discarded items resulting from resize.
+  int resize(int newSize) {
+    _capacity = newSize;
+    return _dropOverflowItems(newSize);
+  }
+}
+
+/// A callback for [ChannelBuffers.drain], called as it pops stored messages.
+typedef DrainChannelCallback = Future<void> Function(ByteData, PlatformMessageResponseCallback);
+
+/// Storage of channel messages until the channels are completely routed,
+/// i.e. when a message handler is attached to the channel on the framework side.
+///
+/// Each channel has a finite buffer capacity and in a FIFO manner messages will
+/// be deleted if the capacity is exceeded.  The intention is that these buffers
+/// will be drained once a callback is setup on the BinaryMessenger in the
+/// Flutter framework.
+///
+/// Clients of Flutter shouldn't need to allocate their own ChannelBuffers
+/// and should only access this package's [channelBuffers] if they are writing
+/// their own custom [BinaryMessenger].
+class ChannelBuffers {
+  /// By default we store one message per channel.  There are tradeoffs associated
+  /// with any size.  The correct size should be chosen for the semantics of your
+  /// channel.
+  ///
+  /// Size 0 implies you want to ignore any message that gets sent before the engine
+  /// is ready (keeping in mind there is no way to know when the engine is ready).
+  ///
+  /// Size 1 implies that you only care about the most recent value.
+  ///
+  /// Size >1 means you want to process every single message and want to chose a
+  /// buffer size that will avoid any overflows.
+  static const int kDefaultBufferSize = 1;
+
+  final Map<String, _RingBuffer<_StoredMessage>> _messages =
+    <String, _RingBuffer<_StoredMessage>>{};
+
+  _RingBuffer<_StoredMessage> _makeRingBuffer(int size) {
+    final _RingBuffer<_StoredMessage> result = _RingBuffer<_StoredMessage>(size);
+    result.dropItemCallback = _onDropItem;
+    return result;
+  }
+
+  void _onDropItem(_StoredMessage message) {
+    message.callback(null);
+  }
+
+  /// Returns true on overflow.
+  bool push(String channel, ByteData data, PlatformMessageResponseCallback callback) {
+    _RingBuffer<_StoredMessage> queue = _messages[channel];
+    if (queue == null) {
+      queue = _makeRingBuffer(kDefaultBufferSize);
+      _messages[channel] = queue;
+    }
+    final bool result = queue.push(_StoredMessage(data, callback));
```

Comment:
just call this overflowed locally if that's what it's supposed to mean

Reply:
Done.